### ISSUE TITLE
Race condition is possible on submitting solution to exercise

### ISSUE THREAD

**Issue by julian-berbel** [mumuki/mumuki-laboratory#1696]
We're not sure how, but a user submitted a solution to an exercise twice simultaneously.
This caused two assignments to be created for the same `(user + exercise + organization)` combination, which caused strange behavior for the user, since the guide progress list would show the exercise as failed, but the results for the exercise would should it as passed (after the solution was fixed by the user).

It seems to me there are two problems mixed here:
- the guide progress list is not guaranteed to get the same assignment as the exercise view, which is kinda weird,
- and the race condition itself; ideally it should not be possible to create two assignments for the same `(user + exercise + organization)` combination.

**Comment by julian-berbel** [comment]
transaction should probably go here: https://github.com/mumuki/mumuki-domain/blob/43e1946e3c52d415761ed05485541d71e06b39c3/app/models/concerns/with_assignments.rb#L31

**Comment by flbulgarelli** [comment]
Yup, I agree. However that _may_ introduce a performance  hit, since that code is used frequently. 

Another solution may be adding a unique constraint in the database for `(user_id, organization_id, exercise_id)`. We might even handle it and find for object again.

**Comment by flbulgarelli** [comment]
:up:  Oh, no, we can not use a transaction there since `build` just instantiates the object, but it does not persist it, so adding a transaction there will not make it thread-safe.

We definitely should go for the unique constraint. Error handling should partially go to controller.
